escrow/src/main.sw

```diff
@@ -1,257 +1,346 @@
+        let escrow = storage.escrows.get(identifier);
 
-                transfer_to_output(storage.asset_amount, storage.asset, storage.seller.address);
-            }
-        } else {
-            revert(0);
-        };
+        require(escrow.state == State::Pending, StateError::StateNotPending);
+
+        let user = msg_sender().unwrap();
+
+        require(user == escrow.seller.address, UserError::Unauthorized);
+        require(arbiter.address != escrow.buyer.address, ArbiterInputError::CannotBeBuyer);
+        require(arbiter.address != escrow.seller.address, ArbiterInputError::CannotBeSeller);
+        require(0 < arbiter.fee_amount, ArbiterInputError::FeeCannotBeZero);
+        require(arbiter.fee_amount == msg_amount(), ArbiterInputError::FeeDoesNotMatchAmountSent);
+        require(arbiter.asset == msg_asset_id(), ArbiterInputError::AssetDoesNotMatch);
+
+        // If there is a previous proposal then we must transfer those funds back to the seller
+        let proposal = storage.arbiter_proposal.get(identifier);
+        if proposal.is_some() && 0 < proposal.unwrap().fee_amount {
+            transfer(proposal.unwrap().fee_amount, proposal.unwrap().asset, escrow.seller.address);
+        }
 
-        true
+        storage.arbiter_proposal.insert(identifier, Option::Some(arbiter));
+
+        log(ProposedArbiterEvent {
+            arbiter, identifier
+        });
+    }
+
+    #[storage(read, write)]fn resolve_dispute(identifier: u64, payment_amount: u64, user: Identity) {
+        // The assertions ensure that a resolution can only occur during a dispute and only once
+        // by the specified arbiter. The deposit will be sent to either the buyer or seller and the
+        // arbiter can choose their payment amount up to the deposit from the seller
+
+        let mut escrow = storage.escrows.get(identifier);
+
+        require(escrow.state == State::Pending, StateError::StateNotPending);
+        require(escrow.disputed, StateError::NotDisputed);
+        require(msg_sender().unwrap() == escrow.arbiter.address, UserError::Unauthorized);
+        require(user == escrow.buyer.address || user == escrow.seller.address, UserInputError::InvalidRecipient);
+        require(escrow.buyer.asset.is_some() && 0 < escrow.buyer.deposited_amount, StateError::CannotResolveBeforeDesposit);
+        require(payment_amount <= escrow.arbiter.fee_amount, ArbiterInputError::PaymentTooLarge);
+
+        escrow.state = State::Completed;
+        storage.escrows.insert(identifier, escrow);
+
+        transfer(escrow.buyer.deposited_amount, escrow.buyer.asset.unwrap(), user);
+        transfer(payment_amount, escrow.arbiter.asset, escrow.arbiter.address);
+
+        if payment_amount != escrow.arbiter.fee_amount {
+            transfer(escrow.arbiter.fee_amount - payment_amount, escrow.arbiter.asset, escrow.seller.address);
+        }
+
+        // If there is a previous proposal then we must transfer those funds back to the seller
```

Comment:
It could be but the issue is the storage which updates the proposal. That cannot be done in a library therefore I would need to duplicate the if conditional check multiple times in order to update the proposal at which point it's probably the same amount of lines anyway so I do not see the benefit